File: server/src/services/pricing.ts
import { getDb } from '../db/index.js';
import * as schema from '../db/schema.js';
import { eq, and } from 'drizzle-orm';

export interface CostEstimate {
  promptCost: number;
  completionCost: number;
  totalCost: number;
  currency: string;
  pricingPrompt: number | null;
  pricingCompletion: number | null;
}

export function calculateCost(
  platform: string,
  modelId: string,
  promptTokens: number,
  completionTokens: number,
): CostEstimate {
  const db = getDb();
  const model = db.select({
    pricingPrompt: schema.models.pricingPrompt,
    pricingCompletion: schema.models.pricingCompletion,
  })
    .from(schema.models)
    .where(and(eq(schema.models.platform, platform), eq(schema.models.modelId, modelId)))
    .get();

  const pricingPrompt = model?.pricingPrompt ?? null;
  const pricingCompletion = model?.pricingCompletion ?? null;

  if (pricingPrompt == null && pricingCompletion == null) {
    return { promptCost: 0, completionCost: 0, totalCost: 0, currency: 'USD', pricingPrompt: null, pricingCompletion: null };
  }

  const promptCost = promptTokens * (pricingPrompt ?? 0);
  const completionCost = completionTokens * (pricingCompletion ?? 0);

  return {
    promptCost: Math.round(promptCost * 1_000_000) / 1_000_000,
    completionCost: Math.round(completionCost * 1_000_000) / 1_000_000,
    totalCost: Math.round((promptCost + completionCost) * 1_000_000) / 1_000_000,
    currency: 'USD',
    pricingPrompt,
    pricingCompletion,
  };
}

export function getAllPricing(): Array<{
  platform: string;
  modelId: string;
  displayName: string;
  pricingPrompt: number | null;
  pricingCompletion: number | null;
  freeTier: boolean;
}> {
  const db = getDb();
  return db.select({
    platform: schema.models.platform,
    modelId: schema.models.modelId,
    displayName: schema.models.displayName,
    pricingPrompt: schema.models.pricingPrompt,
    pricingCompletion: schema.models.pricingCompletion,
    freeTier: schema.models.freeTier,
  })
    .from(schema.models)
    .where(eq(schema.models.enabled, 1))
    .all()
    .map(m => ({
      ...m,
      freeTier: m.freeTier === 1,
    }));
}

File: crates/server/src/services/pricing.rs
//! Model pricing lookup and cost calculation.

use rusqlite::Connection;
use serde::Serialize;

#[derive(Serialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct CostEstimate {
    pub prompt_cost: f64,
    pub completion_cost: f64,
    pub total_cost: f64,
    pub currency: String,
    pub pricing_prompt: Option<f64>,
    pub pricing_completion: Option<f64>,
}

fn round6(v: f64) -> f64 {
    (v * 1_000_000.0).round() / 1_000_000.0
}

pub fn calculate_cost(
    conn: &Connection,
    platform: &str,
    model_id: &str,
    prompt_tokens: i64,
    completion_tokens: i64,
) -> CostEstimate {
    let row: Option<(Option<f64>, Option<f64>)> = conn
        .query_row(
            "SELECT pricing_prompt, pricing_completion FROM models \
             WHERE platform = ?1 AND model_id = ?2",
            rusqlite::params![platform, model_id],
            |r| Ok((r.get(0)?, r.get(1)?)),
        )
        .ok();

    let pricing_prompt = row.and_then(|r| r.0);
    let pricing_completion = row.and_then(|r| r.1);

    if pricing_prompt.is_none() && pricing_completion.is_none() {
        return CostEstimate {
            prompt_cost: 0.0,
            completion_cost: 0.0,
            total_cost: 0.0,
            currency: "USD".to_string(),
            pricing_prompt: None,
            pricing_completion: None,
        };
    }

    let prompt_cost = prompt_tokens as f64 * pricing_prompt.unwrap_or(0.0);
    let completion_cost = completion_tokens as f64 * pricing_completion.unwrap_or(0.0);

    CostEstimate {
        prompt_cost: round6(prompt_cost),
        completion_cost: round6(completion_cost),
        total_cost: round6(prompt_cost + completion_cost),
        currency: "USD".to_string(),
        pricing_prompt,
        pricing_completion,
    }
}

#[derive(Serialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct PricingRow {
    pub platform: String,
    pub model_id: String,
    pub display_name: String,
    pub pricing_prompt: Option<f64>,
    pub pricing_completion: Option<f64>,
    pub free_tier: bool,
}

pub fn get_all_pricing(conn: &Connection) -> Vec<PricingRow> {
    let mut stmt = conn
        .prepare(
            "SELECT platform, model_id, display_name, pricing_prompt, pricing_completion, free_tier \
             FROM models WHERE enabled = 1",
        )
        .unwrap();
    stmt.query_map([], |r| {
        Ok(PricingRow {
            platform: r.get(0)?,
            model_id: r.get(1)?,
            display_name: r.get(2)?,
            pricing_prompt: r.get(3)?,
            pricing_completion: r.get(4)?,
            free_tier: r.get::<_, i64>(5)? == 1,
        })
    })
    .unwrap()
    .filter_map(|r| r.ok())
    .collect()
}
